File: references/python/aml_adverse_media_search.py
from pymongo import MongoClient
from typing import List, Dict, Any
import os
import json
from utilities.text_similarity import get_similarities  # You'll need an equivalent Python package

DATASET_MONGO_URI = os.environ.get('DATASET_MONGO_URI')
DATASET_DB_NAME = os.environ.get('DATASET_DB_NAME')

cached_client = None
cached_db = None

def get_mongo_client():
    global cached_client, cached_db
    
    if cached_client and cached_db:
        try:
            cached_db.command('ping')
            return cached_client
        except Exception:
            cached_client.close()
            cached_client = None
            cached_db = None
    
    try:
        client = MongoClient(DATASET_MONGO_URI)
        cached_client = client
        cached_db = client[DATASET_DB_NAME]
        return cached_client
    except Exception as error:
        print(f'Failed to connect to MongoDB: {error}')
        raise Exception('Failed to connect to MongoDB')

async def get_adverse_media(db, name: str) -> List[Dict]:
    # Preprocess search name
    normalized_search_name = name.lower().strip()
    print(f'Original search name: {name}')
    print(f'Normalized search name: {normalized_search_name}')

    # Get candidates with loose conditions
    pipeline = [
        # Unwind target array
        {"$unwind": {"path": "$target", "preserveNullAndEmptyArrays": True}},
        
        # Match conditions
        {
            "$match": {
                "$or": [
                    {"target.name_en": {"$regex": f".*{normalized_search_name}.*", "$options": "i"}},
                    {"target.name_zh": {"$regex": f".*{normalized_search_name}.*", "$options": "i"}},
                    {"target.en.ceName": {"$regex": f".*{normalized_search_name}.*", "$options": "i"}},
                    {"target.zh.ceName": {"$regex": f".*{normalized_search_name}.*", "$options": "i"}},
                ]
            }
        },
        
        # Regroup documents with same _id
        {
            "$group": {
                "_id": "$_id",
                "doc": {"$first": "$$ROOT"}
            }
        },
        
        # Restore document structure
        {"$replaceRoot": {"newRoot": "$doc"}}
    ]

    print(f'Aggregation pipeline: {json.dumps(pipeline, indent=2)}')
    
    candidates = list(db.adverse_media.aggregate(pipeline))
    print(f'Found {len(candidates)} candidate records')

    # Filter using string similarity
    result = []
    for item in candidates:
        print(f'Checking item: {json.dumps(item.get("target"), indent=2)}')

        if item.get('target', {}).get('name_en'):
            print('Comparing strings:', {
                'search_name': normalized_search_name,
                'target_name': item['target']['name_en'].lower(),
            })
            
            similarities = get_similarities(
                normalized_search_name,
                [item['target']['name_en'].lower()],
                case_sensitive=False,
                order_sensitive=False,
                threshold=20,
                threshold_type='>='
            )
            
            print(f'Similarity results: {similarities}')
            
            if similarities:
                print(f'Found match! Similarity: {similarities[0]}')
                result.append(item)
                continue

        # Check second format
        if item.get('target', {}).get('en'):
            print(f'Checking second format: {json.dumps(item["target"]["en"], indent=2)}')

            for target in item['target']['en']:
                print('Comparing strings:', {
                    'search_name': normalized_search_name,
                    'target_name': target['ceName'].lower(),
                })
                
                similarities = get_similarities(
                    normalized_search_name,
                    [target['ceName'].lower()],
                    case_sensitive=False,
                    order_sensitive=False,
                    threshold=20,
                    threshold_type='>='
                )
                
                if similarities:
                    print(f'Found match! Similarity: {similarities[0]}')
                    result.append(item)
                    break

    print(f'After similarity filtering: {len(result)} records')
    
    formatted_data = []
    for item in result:
        subtitle = item['source']['title']
        if item.get('published'):
            subtitle += f" - {item['published']}"
            
        formatted_data.append({
            '_id': item['_id'],
            'title': item.get('headline', {}).get('en'),
            'link': item['urls'][0] if item.get('urls') else None,
            'subtitle': subtitle,
            'description': item.get('content', {}).get('en'),
        })

    return formatted_data

def handler(event: Dict[str, Any]) -> Dict[str, Any]:
    name_to_search_arr = event.get('nameToSearchArr')

    try:
        client = get_mongo_client()
        db = client[DATASET_DB_NAME]
        
        data = []
        for name in name_to_search_arr:
            try:
                results = get_adverse_media(db, name)
                data.extend(results)
            except Exception as error:
                print(f'Error searching for adverse media with name {name}: {error}')
                continue

        return {
            'statusCode': 200,
            'body': json.dumps({
                'status': 200,
                'message': 'Success',
                'data': data
            })
        }
    except Exception as error:
        print(f'Adverse media search failed: {error}')
        return {
            'statusCode': 500,
            'body': json.dumps({
                'message': 'Adverse media search failed',
                'error': str(error)
            })
        } 

File: references/AMLAdverseMediaSearch.js
const { MongoClient } = require('mongodb');
const { getSimilarities } = require('calculate-string-similarity');

const DATASET_MONGO_URI = process.env.DATASET_MONGO_URI;
const DATASET_DB_NAME = process.env.DATASET_DB_NAME;

let cachedClient = null;
let cachedDb = null;

async function getMongoClient() {
    if (cachedClient && cachedDb) {
        try {
            await cachedDb.command({ ping: 1 });
            return cachedClient;
        } catch (e) {
            await cachedClient.close();
            cachedClient = null;
            cachedDb = null;
        }
    }

    try {
        const client = new MongoClient(DATASET_MONGO_URI);
        await client.connect();
        cachedClient = client;
        cachedDb = client.db(DATASET_DB_NAME);
        return cachedClient;
    } catch (error) {
        console.error('Failed to connect to MongoDB:', error);
        throw new Error('Failed to connect to MongoDB');
    }
}

exports.handler = async (event) => {
    const { nameToSearchArr } = event;

    try {
        const client = await getMongoClient();
        const db = client.db(DATASET_DB_NAME);

        const data = await Promise.all(nameToSearchArr.map(async (name) => {
            try {
                return await getAdverseMedia(db, name);
            } catch (error) {
                console.error(`Error searching for adverse media with name ${name}:`, error);
                return [];
            }
        }));

        const flattenedData = data.flat();

        return {
            statusCode: 200,
            body: JSON.stringify({ status: 200, message: 'Success', data: flattenedData })
        };
    } catch (error) {
        console.error('Adverse media search failed:', error);
        return {
            statusCode: 500,
            body: JSON.stringify({ message: 'Adverse media search failed', error: error.message })
        };
    }
};

async function getAdverseMedia(db, name) {
    // 預處理搜尋名稱
    const normalizedSearchName = name.toLowerCase().trim();
    console.log('原始搜尋名稱:', name);
    console.log('正規化後的搜尋名稱:', normalizedSearchName);

    // 先用寬鬆的條件從資料庫取得候選資料
    const pipeline = [
        // 展開 target 陣列
        { $unwind: { path: "$target", preserveNullAndEmptyArrays: true } },

        // 匹配條件
        {
            $match: {
                $or: [
                    // 第一種格式：直接匹配 target.name_en
                    { "target.name_en": { $regex: `.*${normalizedSearchName}.*`, $options: "i" } },
                    { "target.name_zh": { $regex: `.*${normalizedSearchName}.*`, $options: "i" } },

                    // 第二種格式：匹配 target.en.ceName
                    { "target.en.ceName": { $regex: `.*${normalizedSearchName}.*`, $options: "i" } },
                    { "target.zh.ceName": { $regex: `.*${normalizedSearchName}.*`, $options: "i" } },
                ]
            }
        },

        // 重新組合相同文件的不同 target
        {
            $group: {
                _id: "$_id",
                doc: { $first: "$$ROOT" }
            }
        },

        // 恢復文件結構
        { $replaceRoot: { newRoot: "$doc" } }
    ];

    console.log('聚合管道:', JSON.stringify(pipeline, null, 2));

    const candidates = await db.collection("adverse_media").aggregate(pipeline).toArray();
    console.log(`找到 ${candidates.length} 筆候選資料`);

    // 使用字串相似度進行過濾
    const result = candidates.filter(item => {
        console.log('檢查項目:', {
            target結構: JSON.stringify(item.target, null, 2)
        });

        // 檢查 target 物件
        if (item.target?.name_en) {
            console.log('比對字串:', {
                搜尋名稱: normalizedSearchName,
                目標名稱: item.target.name_en.toLowerCase(),
            });

            const similarities = getSimilarities(normalizedSearchName, [item.target.name_en.toLowerCase()], {
                caseSensitive: false,
                orderSensitive: false,
                threshold: 20,
                thresholdType: '>='
            });

            console.log('相似度結果:', similarities);

            if (similarities.length > 0) {
                console.log('找到匹配！相似度:', similarities[0]);
                return true;
            }
            console.log('相似度未達閾值 20%');
            return false;
        }

        // 檢查第二種格式
        if (item.target?.en) {
            console.log('檢查第二種格式:', {
                en結構: JSON.stringify(item.target.en, null, 2)
            });

            const hasMatch = item.target.en.some(target => {
                console.log('比對字串:', {
                    搜尋名稱: normalizedSearchName,
                    目標名稱: target.ceName.toLowerCase(),
                });

                const similarities = getSimilarities(normalizedSearchName, [target.ceName.toLowerCase()], {
                    caseSensitive: false,
                    orderSensitive: false,
                    threshold: 20,
                    thresholdType: '>='
                });

                console.log('相似度結果:', similarities);

                if (similarities.length > 0) {
                    console.log('找到匹配！相似度:', similarities[0]);
                    return true;
                }
                console.log('相似度未達閾值 20%');
                return false;
            });

            return hasMatch;
        }

        console.log('未找到匹配');
        return false;
    });

    console.log(`相似度過濾後剩下 ${result.length} 筆資料`);
    let data = [];
    if (result.length > 0) {
        result.map(item => {
            const subtitle = item.source.title + (item.published ? ` - ${item.published}` : '')
            data.push({
                _id: item._id,
                title: item.headline?.en,
                link: item.urls[0],
                subtitle: subtitle,
                description: item.content?.en,
            })

        })
    }

    return result;
} 